src/lsm/manifest_level.zig

```diff
@@ -261,36 +262,51 @@ pub fn ManifestLevelType(
+        /// - If the exact table exists in the ManifestLevel: remove it, and return whether the
+        ///   table is visible to any of the given snapshots or `snapshot_latest`.
+        /// The `table` parameter must *not* be a pointer into the `tables`' SegmentedArray memory.
+        pub fn remove_table(
             level: *Self,
             node_pool: *NodePool,
+            snapshots: []const u64,
             table: *const TableInfo,
-        ) *const TableInfo {
-            assert(table.visible(lsm.snapshot_latest));
+        ) ?Visibility {
+            assert(level.keys.len() == level.tables.len());
+            assert(compare_keys(table.key_min, table.key_max) != .gt);
 
-            level.remove_table(node_pool, table);
-            level.table_count_visible -= 1;
-            level.key_range.exclude(KeyRange{
-                .key_min = table.key_min,
-                .key_max = table.key_max,
-            });
+            // Use `key_min` for both ends of the iterator; we are looking for a single table.
+            const cursor_start =
+                level.iterator_start(table.key_min, table.key_min, .ascending) orelse return null;
 
-            return table;
-        }
+            var i = level.keys.absolute_index_for_cursor(cursor_start);
+            var tables = level.tables.iterator_from_index(i, .ascending);
+            const table_index_absolute = while (tables.next()) |level_table| : (i += 1) {
+                // The `table` parameter should *not* be a pointer into the `tables` SegmentedArray
+                // memory, since it will be invalidated by `tables.remove_elements()`.
+                assert(level_table != table);
 
-        /// Remove the given table from the ManifestLevel, asserting that it is not visible
-        /// by any snapshot in `snapshots` or by `lsm.snapshot_latest`.
-        pub fn remove_table_invisible(
-            level: *Self,
-            node_pool: *NodePool,
-            snapshots: []const u64,
-            table: *const TableInfo,
-        ) void {
-            assert(table.invisible(snapshots));
-            level.remove_table(node_pool, table);
+                if (level_table.equal(table)) break i;
+                assert(level_table.checksum != table.checksum);
+                assert(level_table.address != table.address);
+            } else return null;
+
+            level.generation +%= 1;
+            level.keys.remove_elements(node_pool, table_index_absolute, 1);
+            level.tables.remove_elements(node_pool, table_index_absolute, 1);
+            assert(level.keys.len() == level.tables.len());
+
+            if (table.invisible(snapshots)) {
+                return .invisible;
+            } else {
+                level.table_count_visible -= 1;
+                level.key_range.exclude(.{
+                    .key_min = table.key_min,
+                    .key_max = table.key_max,
+                });
+                return .visible;
+            }
```

Comment:
Very good point.

`table_count_visible` is in a similar boat right now -- I added a TODO to move it up from `ManifestLevel` to `Manifest`, since in that case we actually _do_ care about visibility relative to all persistent snapshots, not just `snapshot_latest`.